java/org/apache/catalina/startup/ContextConfig.java

```diff
@@ -1374,7 +1383,19 @@ protected void webConfig() {
     protected void processClasses(WebXml webXml, Set<WebXml> orderedFragments) {
         // Step 4. Process /WEB-INF/classes for annotations and
         // @HandlesTypes matches
-        Map<String, JavaClassCacheEntry> javaClassCache = new HashMap<>();
+
+        Map<String, JavaClassCacheEntry> javaClassCache;
+
+        if (context.getParent() instanceof Host) {
+           Host host = (Host) context.getParent();
+            Container container = host.getParent();
+            CONCURRENCY_LEVEL = container.getStartStopThreads();
+            javaClassCache = new ConcurrentHashMap<>(DEFAULT_CLASS_CACHE_SIZE, DEFAULT_LOAD_FACTOR,
+                    CONCURRENCY_LEVEL);
+        } else {
+            javaClassCache = new ConcurrentHashMap<>(DEFAULT_CLASS_CACHE_SIZE, DEFAULT_LOAD_FACTOR,
+                    CONCURRENCY_LEVEL);
+        }
```

Comment:
Well, if it's changed to be always a ConcurrentHashMap, it probably won't be a measurable performance difference.

Reply:
Have added the default behavior without the flag in the latest commit.